0.12 New Error: too many items for attribute

### Terraform Version

```
# go.mod
github.com/hashicorp/terraform v0.12.0-alpha4.0.20190124012834-1592b4aa67ad
```

### Terraform Configuration Files

```hcl
resource "aws_codedeploy_app" "foo_app" {
  name = "foo-app-%s"
}

resource "aws_iam_role_policy" "foo_policy" {
  name = "foo-policy-%s"
  role = "${aws_iam_role.foo_role.id}"
  policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "autoscaling:CompleteLifecycleAction",
        "autoscaling:DeleteLifecycleHook",
        "autoscaling:DescribeAutoScalingGroups",
        "autoscaling:DescribeLifecycleHooks",
        "autoscaling:PutLifecycleHook",
        "autoscaling:RecordLifecycleActionHeartbeat",
        "codedeploy:*",
        "ec2:DescribeInstances",
        "ec2:DescribeInstanceStatus",
        "tag:GetTags",
        "tag:GetResources",
        "sns:Publish"
      ],
      "Resource": "*"
    }
  ]
}
EOF
}

resource "aws_iam_role" "foo_role" {
  name = "foo-role-%s"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "Service": "codedeploy.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF
}

resource "aws_codedeploy_deployment_group" "foo_group" {
  app_name = "${aws_codedeploy_app.foo_app.name}"
  deployment_group_name = "foo-group-%s"
  service_role_arn = "${aws_iam_role.foo_role.arn}"

  load_balancer_info {
    elb_info {
      name = "foo-elb"
    }
  }
}
```

### Debug Output

OSS TeamCity Links:
* `TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_create`: https://HOSTNAME/repository/download/Aws_ProviderAws/42586:id/debug-aws-78c6ea3-TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_create.log
* `TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithTags`: https://HOSTNAME/repository/download/Aws_ProviderAws/42586:id/debug-aws-78c6ea3-TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithTags.log

### Expected Behavior

Unexpected error is no longer present and acceptance testing passes as it previously did in Terraform 0.11. 

### Actual Behavior

```
--- FAIL: TestAccAWSCodeDeployDeploymentGroup_blueGreenDeployment_complete (28.55s)
    testing.go:568: Step 1 error: errors during plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_create (38.28s)
    testing.go:568: Step 1 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_delete (17.65s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_update (55.32s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_deploymentStyle_create (24.07s)
    testing.go:568: Step 1 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_deploymentStyle_delete (26.74s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_deploymentStyle_update (53.58s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_in_place_deployment_with_traffic_control_create (22.38s)
    testing.go:568: Step 1 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_in_place_deployment_with_traffic_control_update (19.19s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_targetGroupInfo_create (22.04s)
    testing.go:568: Step 1 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_targetGroupInfo_update (22.58s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_targetGroupInfo_delete (18.88s)
    testing.go:568: Step 0 error: errors during follow-up plan:

        Error: too many items for attribute "load_balancer_info"; cannot have more than 1

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_ElasticsearchConfigUpdates (972.85s)
    testing.go:568: Step 1 error: errors during plan:

        Error: too many items for attribute "elasticsearch_configuration"; cannot have more than 1

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_RedshiftConfigUpdates (671.69s)
    testing.go:568: Step 1 error: errors during plan:

        Error: too many items for attribute "s3_configuration"; cannot have more than 1

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_s3basicWithTags (100.59s)
    testing.go:568: Step 1 error: errors during plan:

        Error: too many items for attribute "s3_configuration"; cannot have more than 1

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_s3ConfigUpdates (100.08s)
    testing.go:568: Step 1 error: errors during plan:

        Error: too many items for attribute "s3_configuration"; cannot have more than 1

--- FAIL: TestAccAWSKinesisFirehoseDeliveryStream_SplunkConfigUpdates (100.07s)
    testing.go:568: Step 1 error: errors during plan:

        Error: too many items for attribute "s3_configuration"; cannot have more than 1
```

### Steps to Reproduce

On an AWS provider branch updated with Terraform 0.12 dependencies:

```
TF_ACC=1 go test ./aws -v -timeout 120m -parallel 20 -run=TestAccAWSCodeDeployDeploymentGroup_loadBalancerInfo_create
```

### Additional Context

The schema for these attributes is `MaxItems: 1`, however the configuration and API response does not have more than 1 item.

### References

* `aws_codedeploy_deployment_group` resource `load_balancer_info` attribute schema: https://github.com/terraform-providers/terraform-provider-aws/blob/42dd99ff44bd55cc018fd6ef9a20e09e14d42db5/aws/resource_aws_codedeploy_deployment_group.go#L211-L301
* `aws_codedeploy_deployment_group` resource `load_balancer_info` flattening function: https://github.com/terraform-providers/terraform-provider-aws/blob/42dd99ff44bd55cc018fd6ef9a20e09e14d42db5/aws/resource_aws_codedeploy_deployment_group.go#L1333-L1345

--- a/builtin/providers/test/resource_list_test.go
+++ b/builtin/providers/test/resource_list_test.go
@@ -20,6 +20,27 @@ resource "test_resource_list" "foo" {
 		string = "a"
 		int = 1
 	}
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"test_resource_list.foo", "list_block.#", "1",
+					),
+					resource.TestCheckResourceAttr(
+						"test_resource_list.foo", "list_block.0.string", "a",
+					),
+					resource.TestCheckResourceAttr(
+						"test_resource_list.foo", "list_block.0.int", "1",
+					),
+				),
+			},
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_list" "foo" {
+	list_block {
+		string = "a"
+		int = 1
+	}
 
 	list_block {
 		string = "b"

--- a/builtin/providers/test/resource_nested_set_test.go
+++ b/builtin/providers/test/resource_nested_set_test.go
@@ -59,16 +59,6 @@ resource "test_resource_nested_set" "foo" {
 
 // the empty type_list must be passed to the provider with 1 nil element
 func TestResourceNestedSet_emptyBlock(t *testing.T) {
-	checkFunc := func(s *terraform.State) error {
-		root := s.ModuleByPath(addrs.RootModuleInstance)
-		res := root.Resources["test_resource_nested_set.foo"]
-		for k, v := range res.Primary.Attributes {
-			if strings.HasPrefix(k, "type_list") && v != "1" {
-				return fmt.Errorf("unexpected set value: %s:%s", k, v)
-			}
-		}
-		return nil
-	}
 	resource.UnitTest(t, resource.TestCase{
 		Providers:    testAccProviders,
 		CheckDestroy: testAccCheckResourceDestroy,
@@ -80,7 +70,9 @@ resource "test_resource_nested_set" "foo" {
 	}
 }
 				`),
-				Check: checkFunc,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_nested_set.foo", "type_list.#", "1"),
+				),
 			},
 		},
 	})

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -713,15 +713,25 @@ func (s *GRPCProviderServer) ApplyResourceChange(_ context.Context, req *proto.A
 		}
 	}
 
-	// strip out non-diffs
-	for k, v := range diff.Attributes {
-		if v.New == v.Old && !v.NewComputed && v.NewExtra == "" {
-			delete(diff.Attributes, k)
-		}
-	}
-
 	if private != nil {
 		diff.Meta = private
+	}
+
+	// We need to turn off any RequiresNew. There could be attributes
+	// without changes in here inserted by helper/schema, but if they have
+	// RequiresNew then the state will will be dropped from the ResourceData.
+	for k := range diff.Attributes {
+		diff.Attributes[k].RequiresNew = false
+	}
+
+	// check that any "removed" attributes actually exist in the prior state, or
+	// helper/schema will confuse itself
+	for k, d := range diff.Attributes {
+		if d.NewRemoved {
+			if _, ok := priorState.Attributes[k]; !ok {
+				delete(diff.Attributes, k)
+			}
+		}
 	}
 
 	newInstanceState, err := s.provider.Apply(info, priorState, diff)

--- a/terraform/diff.go
+++ b/terraform/diff.go
@@ -4,6 +4,7 @@ import (
 	"bufio"
 	"bytes"
 	"fmt"
+	"log"
 	"reflect"
 	"regexp"
 	"sort"
@@ -494,11 +495,12 @@ func (d *InstanceDiff) applyBlockDiff(path []string, attrs map[string]string, sc
 		// we need to find the set of all keys that traverse this block
 		candidateKeys := map[string]bool{}
 		blockKey := blockPrefix + n + "."
+		localBlockPrefix := localPrefix + n + "."
 
 		// we can only trust the diff for sets, since the path changes, so don't
 		// count existing values as candidate keys. If it turns out we're
-		// keeping the attributes, we will check catch it down below with
-		// "keepBlock" after we check the set count.
+		// keeping the attributes, we will catch it down below with "keepBlock"
+		// after we check the set count.
 		if block.Nesting != configschema.NestingSet {
 			for k := range attrs {
 				if strings.HasPrefix(k, blockKey) {
@@ -570,7 +572,7 @@ func (d *InstanceDiff) applyBlockDiff(path []string, attrs map[string]string, sc
 			}
 
 			for attr, v := range newAttrs {
-				result[localPrefix+n+"."+attr] = v
+				result[localBlockPrefix+attr] = v
 			}
 		}
 
@@ -589,24 +591,51 @@ func (d *InstanceDiff) applyBlockDiff(path []string, attrs map[string]string, sc
 				if strings.HasPrefix(k, blockKey) {
 					// we need the key relative to this block, so remove the
 					// entire prefix, then re-insert the block name.
-					localKey := n + "." + k[len(blockKey):]
-
+					localKey := localBlockPrefix + k[len(blockKey):]
 					result[localKey] = v
 				}
 			}
 		}
 
-		if countDiff, ok := d.Attributes[strings.Join(append(path, n, ".#"), ".")]; ok {
-
+		if countDiff, ok := d.Attributes[strings.Join(append(path, n, "#"), ".")]; ok {
 			if countDiff.NewComputed {
-				result[localPrefix+n+".#"] = hcl2shim.UnknownVariableValue
+				result[localBlockPrefix+"#"] = hcl2shim.UnknownVariableValue
 			} else {
-				result[localPrefix+n+".#"] = countDiff.New
+				result[localBlockPrefix+"#"] = countDiff.New
+
+				// While sets are complete, list are not, and we may not have all the
+				// information to track removals. If the list was truncated, we need to
+				// remove the extra items from the result.
+				if block.Nesting == configschema.NestingList &&
+					countDiff.New != "" && countDiff.New != hcl2shim.UnknownVariableValue {
+					length, _ := strconv.Atoi(countDiff.New)
+					for k := range result {
+						if !strings.HasPrefix(k, localBlockPrefix) {
+							continue
+						}
+
+						index := k[len(localBlockPrefix):]
+						nextDot := strings.Index(index, ".")
+						if nextDot < 1 {
+							continue
+						}
+						index = index[:nextDot]
+						i, err := strconv.Atoi(index)
+						if err != nil {
+							// this shouldn't happen since we added these
+							// ourself, but make note of it just in case.
+							log.Printf("[ERROR] bas list index in %q: %s", k, err)
+							continue
+						}
+						if i >= length {
+							delete(result, k)
+						}
+					}
+				}
 			}
 		} else {
-			result[localPrefix+n+".#"] = countFlatmapContainerValues(localPrefix+n+".#", result)
+			result[localBlockPrefix+"#"] = countFlatmapContainerValues(localBlockPrefix+"#", result)
 		}
-
 	}
 
 	return result, nil

